DataFrame Documentation (#477)
bodo-ai/Bodo: docs/docs/api_docs/dataframe_lib.md

# Bodo Pandas API (bodo.pandas)

> The bodo.pandas library is designed to scale Pandas workflows with just a one-line change — simply replace:

Reviewer: ```suggestion
The Bodo DataFrame library is designed to accelerate and scale Pandas workflows with just a one-line change — simply replace:
```

``` py
import pandas as pd
```

with

``` py
import bodo.pandas as pd
```

and your existing code can immediately take advantage of high-performance, scalable execution.

Key features include:

- __Full Pandas API compatibility__ with a transparent fallback mechanism to native Pandas, ensuring that your workflows continue uninterrupted even if a feature is not yet supported.

- __Lazy evaluation of operations__, allowing powerful optimizations like filter pushdown and column pruning behind the scenes.

> - __Scalable MPI-based execution__, leveraging HPC (High-Performance Computing) techniques for efficient parallelism; whether you're working on a laptop or running jobs across a large cloud cluster.

Reviewer: ```suggestion
- __Scalable MPI-based execution__, leveraging High-Performance Computing (HPC) techniques for efficient parallelism; whether you're working on a laptop or running jobs across a large cloud cluster.
```

- __Vectorized execution__ with streaming and spill-to-disk capabilities, making it possible to process datasets larger than memory reliably.

!!! warning

>     `bodo.pandas` is under active development and is currently considered experimental. Some features and APIs may not yet be fully supported. We welcome your feedback—please join our community Slack or open an issue on GitHub if you encounter any problems!

Reviewer: ```suggestion
    Bodo DataFrame library is under active development and is currently considered experimental. Some features and APIs may not yet be fully supported. We welcome your feedback—please join our community [Slack](https://bodocommunity.slack.com/join/shared_invite/zt-qwdc8fad-6rZ8a1RmkkJ6eOX1X__knA#/shared-invite/email) or open an issue on [our GitHub](https://github.com/bodo-ai/Bodo) if you encounter any problems!
```

## Lazy Evaluation and Fallback to Pandas

> `bodo.pandas` operates with lazy evaluation, meaning operations are recorded into a query plan rather than executed immediately. Execution is automatically triggered only when results are actually needed, such as when displaying the first 5 rows in a DataFrame with `print(df.head(5))`.

Reviewer: ```suggestion
Bodo DataFrame library operates with lazy evaluation to allow query optimization, meaning operations are recorded into a query plan rather than executed immediately. Execution is automatically triggered only when results are actually needed, such as when displaying the first 5 rows in a DataFrame with `print(df.head(5))`.
```

> If the user code encounters an unsupported Pandas API or an unsupported parameter, `bodo.pandas` gracefully falls back to native Pandas. When this happens, the query plan is immediately executed, the resulting data is collected onto a single core and converted to a Pandas DataFrame, and further operations proceed using Pandas.

Reviewer: ```suggestion
If the user code encounters an unsupported Pandas API or an unsupported parameter, Bodo DataFrame library gracefully falls back to native Pandas. When this happens, the query plan is immediately executed, the resulting data is collected onto a single core and converted to a Pandas DataFrame, and further operations proceed using Pandas.
```

- __Advanced query optimization__ such as
 filter pushdown, column pruning and join reordering behind the scenes.

- __Scalable MPI-based execution__, leveraging High-Performance Computing (HPC) techniques for efficient parallelism;
whether you're working on a laptop or running jobs across a large cloud cluster.

- __Vectorized execution__ with streaming and spill-to-disk capabilities,
making it possible to process datasets larger than memory reliably.

!!! warning
    Bodo DataFrame Library is under active development and is currently considered experimental.
    Some features and APIs may not yet be fully supported.
    We welcome your feedback — please join our community [Slack](https://bodocommunity.slack.com/join/shared_invite/zt-qwdc8fad-6rZ8a1RmkkJ6eOX1X__knA#/shared-invite/email) or open an issue on [our GitHub](https://github.com/bodo-ai/Bodo)
    if you encounter any problems!

## Lazy Evaluation and Fallback to Pandas

Bodo DataFrame Library operates with lazy evaluation to allow query optimization, meaning operations are recorded into a query plan rather than executed immediately.

> Execution is automatically triggered only when results are actually needed, such as when displaying the first 5 rows in a DataFrame with `print(df.head(5))`.

Reviewer: Minor: change `head` example to just printing the dataframe? `head` creates a separate plan and is not the same dataframe being materialized.

If the user code encounters an unsupported Pandas API or an unsupported parameter, Bodo DataFrame library gracefully falls back to native Pandas.

> When this happens, the query plan is immediately executed, the resulting data is collected onto a single core and converted to a Pandas DataFrame, and further operations proceed using Pandas.

Reviewer: ```suggestion
When this happens, the current query plan of the DataFrame is immediately executed, the resulting data is collected onto a single core and converted to a Pandas DataFrame, and further operations proceed using Pandas.
```

!!! warning
    Fallback to Pandas may lead to degraded performance and increase the risk of out-of-memory (OOM) errors, especially for large datasets.

## General Functions

### bodo.pandas.from_pandas

``` py
bodo.pandas.from_pandas(df: pandas.DataFrame) -> BodoDataFrame
```

Converts a Pandas DataFrame into an equivalent BodoDataFrame.

> <p style="font-size: 1.1em; font-weight: bold;">Parameters</p>

Reviewer 2: Do we use these paragraph style tags elsewhere? IMO it's better to just use pure markdown where possible, these could maybe be headers? Not a big deal though
